File: 47주차/전력망둘로나누기.js
function solution(n, wires) {
  var answer = 100; // 두 전력망이 가지는 송전탑 개수 차이 절댓값 구하기.
  let cnt = 0;
  // 배열 초기화.
  var arr = new Array(n + 1);

  for (var i = 0; i < arr.length; i++) {
    arr[i] = new Array(); // n+1 크기로 빈배열 넣기.
  }
  // 노드 연결.
  for (let v of wires) {
    const [a, b] = v;
    arr[a].push(b);
    arr[b].push(a);
  }

  function dfs(start, a, b) {
    visited[start] = 1;
    cnt += 1;
    for (let v of arr[start]) {
      if (
        // a,b번 노드의 연결이 끊어졌을 때,
        !visited[v] &&
        [start, v].toString() !== [a, b].toString() &&
        [v, start].toString() !== [a, b].toString()
      ) {
        visited[v] = 1;
        dfs(v, a, b);
      }
    }
  }

  let visited = new Array(n + 1).fill(0);
  // 하나 끊기
  for (let v of wires) {
    const [a, b] = v;
    let tmp = [];
    // console.log(a,b,"끊기")
    visited = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
      if (!visited[i]) {
        cnt = 0;
        dfs(i, a, b);
        tmp.push(cnt);
      }
    }

    answer = Math.min(answer, Math.abs(tmp[0] - tmp[1]));
  }

  return answer;
}
// 프로그래머스 완전 탐색.

File: 68주차/전력망둘로나누기.py
from collections import deque
def solution(n,wires):
    answer = 1000
    g = [[] for _ in range(n+1)]
    
    for a,b in wires:# 그래프 정보 연결하기.
        g[a].append(b)
        g[b].append(a)
    def bfs(start,no):# start 노드 부터 시작해서 탐색했을 때, 해당 네트워크 송전탑 개수 구하기
        q = deque() # no 번 노드는 연결이 끊김.
        q.append(start)
        visited = [0]*(n+1)
        cnt = 0
        visited[start] =1
        while q:
            x = q.popleft()
            cnt += 1
            for nx in g[x]:
                if not visited[nx] and nx != no:
                    visited[nx] = 1
                    q.append(nx)
        return cnt
    
    for a,b in wires:
        cnt = bfs(a,b)
        answer = min(answer, abs(n -(2*cnt)))# 한 덩어리 개수만 알면 나머지 개수는 n-cnt
    
    return answer